### PR TITLE
Adding fix provider for S1155

### LINKED ISSUE #2334: Feature request: S1155 should have a code fix
### Description
It would be nice if rule S1155 ("Any()" should be used to test for emptiness) has a code fix too.

#### For Fix
```csharp
public bool Zero()
{
    return this.SomeCollection.Count(SomeCodition) == 0;
}
public bool MoreThanZero()
{
    return this.SomeCollection.Count(SomeCodition) > 0;
}
public bool NotZero()
{
    return this.SomeCollection.Count(SomeCodition) != 0;
}
```

#### After Fix

```csharp
public bool Zero()
{
    return !this.SomeCollection.Any(SomeCodition);
}
public bool MoreThanZero()
{
    return this.SomeCollection.Any(SomeCodition);
}
public bool NotZero()
{
    return this.SomeCollection.Any(SomeCodition);
}
```

### PR DESCRIPTION
**Pull request by Corniel**
Fixes #2334. All replacements seem okay, the formatting (and the test) are a bit off. Help is wanted for that part.

### CODE REVIEW

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel ,

Thank you for this PR. There seems to be some conflicts blocking the merge. Can you take a look at that?

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource Rewrote the internals to be able to benefit from the any vs empty conclusion. That was why.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel, test `CollectionEmptinessChecking_CodeFix` failed for CSharp7. Can you take a look?

We do a little bit of magic with language versions during local test runs to save some time and simplify debugging. If you don't see it failing, you can check/remove this for debugging purposes:
https://github.com/SonarSource/sonar-dotnet/blob/074a5a5bc94e00b8755ea9f8b347892d78acdb85/analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/ParseOptionsHelper.cs#L97-L100

`scripts/dev/dev-build.ps1` with `-test` option should also run all language versions.

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource It was failing for all after my FI I noticed. That should be fixed now.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs:36]: This one could remain as single liner

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionEmptinessChecking.Fixed.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionEmptinessChecking.Fixed.cs:19]: Can you add a test case demonstrating that `!` operator will behave nicely with this case?

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs:114]: The empty line should be preserved

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs:95]: To follow our coding style
```suggestion
        private static bool IsMethodCountExtension(IMethodSymbol methodSymbol) =>
            methodSymbol.Name == "Count" 
            && methodSymbol.IsExtensionMethod 
            && methodSymbol.ReceiverType != null;
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessChecking.cs:97]: Can you please reorder members in this struct according to https://github.com/SonarSource/sonar-dotnet/blob/master/docs/coding-style.md ? 

There should be fields, empty line, properties, empty line, constructor, other methods (separated by empty line)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs:40]: Should be single line

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs:85]: `countNode` comes from `FirstOrDefault` but this relies on `First()` part only. Something's not right here.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
First round - I checked only beauty, I didn't take a look at the logic yet, as it's quite complicated.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs:36]: Sure. No guidelines of making lines too long?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionEmptinessChecking.Fixed.cs
Comment (Corniel) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/CollectionEmptinessChecking.Fixed.cs]: I rewrote the tests, using naming to make all scenario's clear. Hop that it now clear which scenario's are covered.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Rules/CollectionEmptinessCheckingTest.cs]: S103 will tell you :) It's configured for 200 characters.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionEmptinessCheckingFixProvider.cs:83]: Did a rebase

**Comment by Corniel** [comment]
@pavel-mikula-sonarsource can you have a look at #4033 I think it is worth to introduce that one first.

**Comment by Corniel** [comment]
Did a rebase.

**Comment by pavel-mikula-sonarsource** [comment]
Hi @Corniel, this one will need a major rebase/rework after #4648 gets merged. 

I'll close this PR for now, as we can't progress on it. Please create a new PR later.

**Pull request closed without merging**
